### PR TITLE
Testerina Implementation 

### LINKED ISSUE #4: Remove Glabal base64EncodedString
https://github.com/wso2-ballerina/package-jira/blob/02e7728e9dfa76a83a16c40dac05d89234741b54/jira/jira_connector.bal#L32

### PR DESCRIPTION
**Pull request by NipunaRanasinghe**
## Purpose
Complete Testerina Implementation for the connector
resolved #4 issue

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: jira/jira_structs.bal [hunk 1]
```diff
@@ -368,42 +59,6 @@ public struct ProjectComponentSummary {
...
-                }
-                ProjectComponent projectComponent => {
-                    return projectComponent;
-                }
-            }
-        }
-    }
-}
-
 public struct ProjectComponent {
```
Comment (muthulee) [inline jira/jira_structs.bal]: Add comments to structs.

File: jira/jira_connector.bal [hunk 2]
```diff
@@ -23,9 +23,9 @@ import ballerina/util;
 import ballerina/log;
 import ballerina/io;
 
+
 //Creates package-global Http client endpoint for jira REST API
-endpoint http:ClientEndpoint jiraHttpClientEP {targets:[{uri:WSO2_STAGING_JIRA_REST_API_ENDPOINT}],
-    chunking:http:Chunking.NEVER};
+endpoint http:ClientEndpoint jiraHttpClientEP {targets:[{uri:WSO2_STAGING_JIRA_REST_API_ENDPOINT}], chunking:http:Chunking.NEVER};
```
Comment (hasithajayasundara) [inline jira/jira_connector.bal]: remove package global variables.

File: README.md [hunk 3]
```diff
@@ -94,16 +95,21 @@ All the actions return two values: result and error. Results can be either`balle
...
     string projectKey = "RRDEVSPRT";
     
+    //Creating the jira Connector as an endpoint
+    endpoint jira:JiraConnectorEndpoint jiraConnectorEP {
+        base_url:"https://support-staging.wso2.com"
+    };
+    
     //Authentication 
-    var response = jiraConnector.authenticate("ashan@wso2.com", "ashan123");
+    var response = jiraConnector -> authenticate("ashan@wso2.com", "ashan123");
```
Comment (muthulee) [inline README.md]: Move authentication to initializer. Why? If I want to login to connector in same runtime using two users they will over ride each other's authentication details.

File: jira/jira_connector_endpoint.bal [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    jira_authentication_ep = jira_base_url + JIRA_AUTH_RESOURCE;
+    jira_rest_api_uri = jira_base_url + JIRA_REST_API_RESOURCE + JIRA_REST_API_VERSION;
+
+}
+
+public function <JiraConnectorEndpoint jiraConnectorEP> register(typedesc serviceType) {
+
+}
+
+public function <JiraConnectorEndpoint jiraConnectorEP> start() {
```
Comment (hasithajayasundara) [inline jira/jira_connector_endpoint.bal]: Add annotations.

Comment (hasithajayasundara) [reply jira/jira_connector_endpoint.bal]: Add annotations.

File: jira/jira_connector.bal [hunk 5]
```diff
@@ -24,11 +24,12 @@ import ballerina/log;
 import ballerina/io;
```
Comment (muthulee) [inline jira/jira_connector.bal]: Licence file?

File: jira/jira_connector_endpoint.bal [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+                                   base_url:jiraConnectorEP.jiraConfig.base_url};
+
+    jira_base_url = jiraConnectorEP.jiraConnector.base_url == "" ?
+                    WSO2_STAGING_JIRA_BASE_URL : jiraConnectorEP.jiraConnector.base_url;
+    jira_authentication_ep = jira_base_url + JIRA_AUTH_RESOURCE;
+    jira_rest_api_uri = jira_base_url + JIRA_REST_API_RESOURCE + JIRA_REST_API_VERSION;
+
+}
+
+public function <JiraConnectorEndpoint jiraConnectorEP> register(typedesc serviceType) {
```
Comment (hasithajayasundara) [inline jira/jira_connector_endpoint.bal]: Add annotations.

File: jira/jira_connector_endpoint.bal [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+import ballerina/io;
+import ballerina/net.http;
+
+public struct JiraConfiguration {
+    http:ClientEndpointConfiguration httpClientConfig;
+    string base_url;
+}
+
+public function <JiraConfiguration jiraConfig> JiraConfiguration() {
```
Comment (hasithajayasundara) [inline jira/jira_connector_endpoint.bal]: Add annotations.

File: jira/jira_connector.bal [hunk 8]
```diff
@@ -24,11 +24,12 @@ import ballerina/log;
...
 
 //Creates package-global Http client endpoint for jira REST API
-endpoint http:ClientEndpoint jiraHttpClientEP {targets:[{uri:WSO2_STAGING_JIRA_REST_API_ENDPOINT}],
-    chunking:http:Chunking.NEVER};
-http:HttpConnectorError connectionError;
+endpoint http:ClientEndpoint jiraHttpClientEP {
```
Comment (muthulee) [inline jira/jira_connector.bal]: Please don't create package Global endpoint.  Endpoint will be created by the end user at the point of using the endpoint.

File: tests/test.bal [hunk 9]
```diff
@@ -1,23 +1,29 @@
...
+import ballerina/net.http;
 import ballerina/io;
 import jira;
 
 jira:JiraConnectorError e = {};
 boolean result;
 boolean isValid;
-jira:JiraConnector jiraConnector = {};
+
+endpoint jira:JiraConnectorEndpoint jiraConnectorEP {base_url:"https://support-staging.wso2.com"};
```
Comment (hasithajayasundara) [inline tests/test.bal]: If possible create a constant for base_url and use it here.

File: tests/test.bal [hunk 10]
```diff
@@ -1,23 +1,29 @@
...
 
 public function main (string[] args) {
```
Comment (hasithajayasundara) [inline tests/test.bal]: Remove main.

File: jira/jira_connector.bal (hunk 5 above)
Comment (NipunaRanasinghe) [reply jira/jira_connector.bal]: It's already there

File: README.md (hunk 3 above)
Comment (muthulee) [reply README.md]: Need to update authetnication.

**Pull request merged**